_overviews/core/architecture-of-scala-213-collections.md

```diff
@@ -0,0 +1,1414 @@
+   operations is given as an appendix.
+
+You have now seen how Scala's collections are built and how you can
+add new kinds of collections. Because of Scala's rich support for
+abstraction, each new collection type has a large number of
+methods without having to reimplement them all over again.
+
+### Acknowledgement ###
+
+These pages contain material adapted from the 2nd edition of
+[Programming in Scala](http://www.artima.com/shop/programming_in_scala) by
+Odersky, Spoon and Venners. We thank Artima for graciously agreeing to its
+publication.
+
+## Appendix: Methods to overload to support the “same result type” principle ##
+
+You want to add overloads to specialize transformation operations such that they return a more specific result type. Examples are:
+- `map`, on `StringOps`, when the mapping function returns a `Char`, should return a `String` (instead of an `IndexedSeq`),
+- `map`, on `Map`, when the mapping function returns a pair, should return a `Map` (instead of an `Iterable`),
+- `map`, on `SortedSet`, when an implicit `Ordering` is available for the resulting element type, should return a
+`SortedSet` (instead of a `Set`).
+
+The following table lists transformation operations that might return an undesirably wide type. You might want to overload
+these operations to return a more specific type.
+
+ Collection    | Operations
+---------------|--------------
+`Iterable`     | `map`, `flatMap`, `collect`, `scanLeft`, `scanRight`, `groupMap`, `concat`, `zip`, `zipAll`, `unzip`
+`Seq`          | `prepended`, `appended`, `prependedAll`, `appendedAll`, `padTo`, `patch`
+`immutable.Seq`| `updated`
+`SortedSet`    | `map`, `flatMap`, `collect`, `zip`
+`Map`          | `map`, `flatMap`, `collect`, `concat`
+`immutable.Map`| `updated`, `transform`
+`SortedMap`    | `map`, `flatMap`, `collect`, `concat`
+`immutable.SortedMap` | `updated`
+
+## Appendix: Cross-building custom collections ##
+
+Since the new internal API of the Scala 2.13 collections is very different from the previous
+collections API, authors of custom collection types should use separate source directories
+(per Scala version) to define them.
+
+With sbt you can achieve this by adding the following setting to your project:
+
+~~~ scala
+// Adds a `src/main/scala-2.13+` source directory for Scala 2.13 and newer
+// and a `src/main/scala-2.13-` source directory for Scala version older than 2.13
+unmanagedSourceDirectories in Compile += {
+  val sourceDir = (sourceDirectory in Compile).value
+  CrossVersion.partialVersion(scalaVersion.value) match {
+    case Some((2, n)) if n >= 13 => sourceDir / "scala-2.13+"
+    case _                       => sourceDir / "scala-2.13-"
+  }
+}
+~~~
+
+And then you can define a Scala 2.13 compatible implementation of your collection
+in the `src/main/scala-2.13+` source directory, and an implementation for the
+previous Scala versions in the `src/main/scala-2.13-` source directory.
```

Comment:
We could include links to projects, like the scalacheck PR (https://github.com/rickynils/scalacheck/pull/411)

Reply:
Done.